src/util/taproot.rs

```diff
@@ -772,23 +776,42 @@ impl LeafVersion {
     // or an opcode that is not valid as the first opcode).
     // The values that comply to this rule are the 32 even values between
     // 0xc0 and 0xfe and also 0x66, 0x7e, 0x80, 0x84, 0x96, 0x98, 0xba, 0xbc, 0xbe
-    pub fn from_u8(ver: u8) -> Result<Self, TaprootError> {
-        if ver & TAPROOT_LEAF_MASK == ver && ver != 0x50 {
-            Ok(LeafVersion(ver))
-        } else {
-            Err(TaprootError::InvalidTaprootLeafVersion(ver))
+    pub fn from_byte(version: u8) -> Result<Self, TaprootError> {
+        match version {
+            TAPROOT_LEAF_TAPSCRIPT => Ok(LeafVersion::TapScript),
+            TAPROOT_ANNEX_PREFIX => Err(TaprootError::InvalidTaprootLeafVersion(TAPROOT_ANNEX_PREFIX)),
+            even if even & TAPROOT_LEAF_MASK != even => Err(TaprootError::InvalidTaprootLeafVersion(even)),
+            future => Ok(LeafVersion::Future(future)),
         }
     }
 
     /// Get the inner version from LeafVersion
-    pub fn as_u8(&self) -> u8 {
-        self.0
+    pub fn as_byte(&self) -> u8 {
+        match self {
+            LeafVersion::TapScript => TAPROOT_LEAF_TAPSCRIPT,
+            LeafVersion::Future(version) => *version,
+        }
     }
 }
 
 impl Into<u8> for LeafVersion {
     fn into(self) -> u8 {
-        self.0
+        self.as_byte()
+    }
+}
+
+impl fmt::Display for LeafVersion {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match (self, f.alternate()) {
+            (LeafVersion::TapScript, false) => f.write_str("tapscript"),
+            (LeafVersion::TapScript, true) => fmt::Display::fmt(&TAPROOT_LEAF_TAPSCRIPT, f),
+            (LeafVersion::Future(version), false) => {
+                f.write_str("future(")?;
+                fmt::Display::fmt(version, f)?;
```

Comment:
I don't think formatter parameters should affect version number. E.g. `"{:10}"` would display `future(.........1)` (spaces instead of dots, GH is broken) instead of `.future(1)`
However, getting parameters right is quite annoying. :(

Reply:
I think we should just always serialize in hex.

Reply:
Hmm, version is the kind of thing that I consider naturally decimal.

Reply:
That's fair.

Ok, "no opinion" on what format it should have :)

Reply:
I agree with @apoelstra that the script "version", which is not ordinal, is better to be formatted as hex: it will be odd to see the next post-tapscript version as `future(82)` than `future(0x52)`. Decimals will do a bad job here.

I agree with @sanket1729 that we should not use formatting when we use version "number" (actually version id) as a part of other string.

I fixed both of the problems and also changed from `future(x)` to `future_script_0xXX` so it will be more readable. When the version is printed without any other string in alternate representation, I pass the formatting flags to allow users to format it as they like (decimal, prefixed with zeros etc).